### PR TITLE
lwm2m_client and utils: Location assistance object and usage

### PR DESCRIPTION
**Pull request by jarlamsa**
Provide location assistance object that is used in accordance with coiote lwm2m-server.
Currently the object supports 2 types of location assistance and is working with the sample lwm2m_client.
* Cell based location assistance (Single cell and multi cell)
* A-GPS based location assistance
Both of these assistance types are being forwarded to the nrfcloud by coiote lwm2m-server.

### CODE REVIEW

File: samples/nrf9160/lwm2m_client/overlay-avsystem-agps.conf [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+CONFIG_LOCATION=y
+CONFIG_NRF_CLOUD_AGPS=y
+CONFIG_LOCATION_METHOD_CELLULAR=n
+CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL=y
+CONFIG_GNSS_ANTENNA_EXTERNAL=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS=y
+CONFIG_LWM2M_VERSION_1_1=y
+CONFIG_BASE64=y
+CONFIG_LWM2M_RW_SENML_JSON_SUPPORT=y
+CONFIG_LWM2M_RW_JSON_SUPPORT=y
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/overlay-avsystem-agps.conf:11]: Actually Seppo have create already PR for lwm2m which include some of the configs.

**Review by jheiskan81 — COMMENTED** [review]
Looks good for me. Those overlay configures may need some action.
https://github.com/nrfconnect/sdk-nrf/pull/7198

**Comment by jarlamsa** [comment]
Thanks @simensrostad thank you for your review. I think I have addressed the issues you mentioned. Please feel free to re-review.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+static int32_t pgps_pred_interval;
+static int32_t pgps_start_gps_day;
+static int32_t pgps_start_gps_time_of_day;
+//Assistance data
+static char assist_data[1024];
+static int32_t result;
+static float latitude;
+static float longitude;
+static float altitude;
+static float accuracy;
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c:66]: We should use `double` for those - a bit counter-intuitive but engine represents LwM2M float type as double for best accuracy.

File: subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#include <net/lwm2m_path.h>
+#include <net/lwm2m_client_utils_location.h>
+#include <net/lwm2m_client_utils.h>
+
+#define MODULE lwm2m_location_event_handler
+
+static struct lwm2m_ctx *client_ctx;
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_LWM2M_LOG_LEVEL);
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c:25]: Utils lib has its own log level (CONFIG_LWM2M_CLIENT_UTILS_LOG_LEVEL).

File: subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ */
+
+#include <zephyr.h>
+#include <nrf_modem_gnss.h>
+
+#include <net/lwm2m_path.h>
+#include <net/lwm2m_client_utils_location.h>
+#include <net/lwm2m_client_utils.h>
+
+#define MODULE lwm2m_location_event_handler
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c:14]: Do we need this define? Could use the module name directly in `LOG_MODULE_REGISTER()`

File: subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c [hunk 5]
```diff
@@ -54,269 +54,14 @@ static int32_t pgps_start_gps_day;
 static int32_t pgps_start_gps_time_of_day;
 //Assistance data
 static char assist_buf[4096];
-
-static char rawData[3023] = {
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c]: I think the commits could be cleaned up a bit, there's really no point introducing changes in a PR just to remove them in the next commit.

File: subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+		mask |= LOCATION_ASSIST_NEEDS_LOCATION;
+	}
+	if (event->agps_req.data_flags & NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) {
+		mask |= LOCATION_ASSIST_NEEDS_INTEGRITY;
+	}
+
+	location_assist_agps_request_set(mask);
+
+	char const *send_path[7] = {
+	LWM2M_PATH(LOCATION_ASSIST_OBJECT_ID, 0, LOCATION_ASSIST_ASSIST_TYPE),
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c:62]: The block content should be indented one tab more.

File: subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	if (send_back) {
+		LOG_INF("Cell location event with send back request");
+		location_assist_cell_request_set();
+	} else {
+		LOG_INF("Cell location inform event");
+		location_assist_cell_inform_set();
+	}
+
+	char const *send_path[CELL_LOCATION_PATHS] = {
+	LWM2M_PATH(LOCATION_ASSIST_OBJECT_ID, 0, LOCATION_ASSIST_ASSIST_TYPE),
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/location/location_event_handler.c:91]: The block content should be indented one tab more.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c (hunk 5 above)
Comment (jarlamsa) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c]: Yeah, I agree. That is something that I have been working on and still there's work to do. At least in this file.

**Comment by jarlamsa** [comment]
Rebased and cleaned up the commits.

**Review by rlubos — APPROVED** [review]

**Review by jheiskan81 — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

File: doc/nrf/releases/release-notes-changelog.rst [hunk 8]
```diff
@@ -221,6 +221,7 @@ nrf9160 Samples
       Added :ref:`state_diagram` to the sample documentation.
     * Replaced the deprecated GPS-driver with the new GNSS interface.
     * Added minimal Portfolio object support that is required for LwM2M conformance testing.
+    * Added support for using location assistance when using Coiote LwM2M-server.
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:224]:
```suggestion
    * Added support for using location assistance when using Coiote LwM2M server.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 9]
```diff
@@ -365,6 +366,7 @@ Libraries for networking
   * :ref:`lib_lwm2m_client_utils` library:
 
     * Updated the library to store credentials and server settings permanently on bootstrap.
+    * Added support for using location assistance when using Coiote LwM2M-server.
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:369]:
```suggestion
    * Added support for using location assistance when using Coiote LwM2M server.
```

**Comment by jarlamsa** [comment]
Thank you @SeppoTakalo and @umapraseeda for your review comments. I think I have addressed those. @umapraseeda I recently separated the readme now to be used as a subpage in the future, to `location_assistance.rst`, could you also review that?

File: samples/nrf9160/lwm2m_client/location_assistance.rst [hunk 10]
```diff
@@ -0,0 +1,34 @@
+.. _location_assistance:
+
+Location Assistance
+===================
+
+The sample supports location assistance from nRF Cloud. This is achieved with using lwm2m-objects Signal Measurement Information (ID 10256) and Location assistance (ID 50001).
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/location_assistance.rst:6]:
```suggestion
The sample supports proprietary mechanism to fetch location assistance data from nRF Cloud by proxying it through LwM2M. This is achieved with using lwm2m-objects Signal Measurement Information (ID 10256) and proprietary Location assistance object (ID 50001).
```

File: samples/nrf9160/lwm2m_client/location_assistance.rst [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+Currently the Coiote lwm2m-server can be used for utilizing the location assistance with nRF Cloud.
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/location_assistance.rst:7]:
```suggestion
This feature is currently under development and right now, only the AVSystem Coiote lwm2m-server can be used for utilizing the location assistance with nRF Cloud.
```

File: samples/nrf9160/lwm2m_client/location_assistance.rst [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+There are currently two different methods of obtaining the location assistance.
+
+First is the location based on cell information. The device sends the information about the current cell and possibly also neighbouring cells to the lwm2m-server.
+The lwm2m-server then sends the location request to the nRF Cloud which responds with the location.
+
+Second method is requesting A-GPS assistance data, which is queried from the nRF Cloud and fed through the lwm2m-server back to the device. The A-GPS assistance data is then fed to the GNSS-module for obtaining a location faster.
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/location_assistance.rst:14]:
```suggestion
Second method is requesting A-GPS assistance data, which is queried from the nRF Cloud and fed through the lwm2m-server back to the device. The A-GPS assistance data is then fed to the GNSS-module for obtaining a position fix faster.
```

**Review by SeppoTakalo — APPROVED** [review]
Looks good.

Added few proposals for documentation

File: samples/nrf9160/lwm2m_client/location_assistance.rst (hunk 11 above)
Comment (umapraseeda) [reply samples/nrf9160/lwm2m_client/location_assistance.rst]: Made this a note.

File: samples/nrf9160/lwm2m_client/location_assistance.rst (hunk 12 above)
Comment (umapraseeda) [reply samples/nrf9160/lwm2m_client/location_assistance.rst]: Updated

File: doc/nrf/libraries/networking/lwm2m_client_utils.rst [hunk 13]
```diff
@@ -18,6 +18,8 @@ Following are the fixed set of readily initialized objects that are available to
 * Device object
 * Location object
 * LwM2M Security object
+* Signal measurement information object
```
Comment (umapraseeda) [inline doc/nrf/libraries/networking/lwm2m_client_utils.rst:21]:
```suggestion
* Signal Measurement Information object
```

File: doc/nrf/libraries/networking/lwm2m_client_utils.rst [hunk 14]
```diff
@@ -40,12 +42,17 @@ Enable the following parameters when using this library:
 * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT`
 * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_OBJ_SUPPORT`
 * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT`
+* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_OBJ_SUPPORT`
+* :kconfig:option:`LWM2M_CLIENT_UTILS_LOCATION_ASSIST_OBJ_SUPPORT`
```
Comment (umapraseeda) [inline doc/nrf/libraries/networking/lwm2m_client_utils.rst:46]:
```suggestion
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_OBJ_SUPPORT`
```

File: doc/nrf/libraries/networking/lwm2m_client_utils.rst [hunk 15]
```diff
@@ -40,12 +42,17 @@ Enable the following parameters when using this library:
...
 
 Support for the objects are enabled by default but they can be set individually.
 Disable the :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_DEVICE_OBJ_SUPPORT` Kconfig option only if you are implementing a ``Reboot`` resource on your application because of a mandatory requirement.
 
 If you are using the Firmware Update object and require downloading of firmware images from TLS enabled services like HTTPS, configure :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_DOWNLOADER_SEC_TAG` to specify the security tag that has root certificate for the target server.
 
+Location assistance object is a proprietary LwM2M object used to deliver information required by various location services through LwM2M.
+It is still under development and considered experimental. Please see :ref:`location_assistance` section on LwM2M client for more information.
```
Comment (umapraseeda) [inline doc/nrf/libraries/networking/lwm2m_client_utils.rst:54]:
```suggestion
Location Assistance object is a proprietary LwM2M object used to deliver information required by various location services through LwM2M.
It is still under development and considered experimental.
See the :ref:`location_assistance` section in the LwM2M client sample documentation for more information.
```

File: doc/nrf/releases/release-notes-changelog.rst [hunk 16]
```diff
@@ -469,6 +471,7 @@ Libraries for Zigbee
   * Extended ``zcl cmd`` shell command to allow sending groupcasts.
   * Extended ``zdo`` shell commands to allow binding to a group addresses.
   * Fixed an issue where printing binding table containing group-binding entries results in corrupted output.
+  * Aded a set of ``zcl groups`` shell commands for managing Zigbee groups.
```
Comment (umapraseeda) [inline doc/nrf/releases/release-notes-changelog.rst:474]: Duplicate entry added. Please remove this. Also https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/releases/release-notes-changelog.html some more entries here. So may be some merge conflicts need to be resolved to update the file.

File: samples/nrf9160/lwm2m_client/location_assistance.rst [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  Configuration for using A-GPS as location assistance method
+  nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
+
+* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS`
+  Configuration for using events from :ref:`lib_lwm2m_client_utils` library for sending assistance requests
+  Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
+
+* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER`
+  Configuration for using a listener from :ref:`lib_lwm2m_client_utils` library for populating the cell neighborhood
+  Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.
```
Comment (umapraseeda) [inline samples/nrf9160/lwm2m_client/location_assistance.rst:37]: Suggesting to remove the descriptions ( like 'Configuration for using cell based location as a location assistance method.') that are already present in the Kconfig description.
```suggestion
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` - nRF Cloud estimates the location of the device based on the cell neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS` -  nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER` - Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.
```

Comment (umapraseeda) [inline samples/nrf9160/lwm2m_client/location_assistance.rst:37]: Suggesting to remove the descriptions ( like 'Configuration for using cell based location as a location assistance method.') that are already present in the Kconfig description.
```suggestion
Check and configure the following library options that are used by the sample:

* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` - nRF Cloud estimates the location of the device based on the cell neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS` -  nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER` - Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.
```

File: doc/nrf/libraries/networking/lwm2m_client_utils.rst (hunk 14 above)
Comment (SeppoTakalo) [reply doc/nrf/libraries/networking/lwm2m_client_utils.rst]: fixed.

File: doc/nrf/releases/release-notes-changelog.rst (hunk 16 above)
Comment (SeppoTakalo) [reply doc/nrf/releases/release-notes-changelog.rst]: Dropped this line. Probably conflict resolving gone wrong on my end.

**Review by umapraseeda — APPROVED** [review]
LGTM! Thanks!

File: samples/nrf9160/lwm2m_client/overlay-assist-agps.conf [hunk 18]
```diff
@@ -0,0 +1,12 @@
+CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_OBJ_SUPPORT=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS=y
+CONFIG_NRF_CLOUD_AGPS=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS=y
+CONFIG_LWM2M_VERSION_1_1=y
+CONFIG_BASE64=y
+CONFIG_LWM2M_RW_SENML_JSON_SUPPORT=y
+CONFIG_LWM2M_RW_JSON_SUPPORT=y
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/overlay-assist-agps.conf:9]: Is there a need for the older JSON format?

File: samples/nrf9160/lwm2m_client/overlay-assist-cell.conf [hunk 19]
```diff
@@ -0,0 +1,11 @@
+CONFIG_NEWLIB_LIBC_FLOAT_PRINTF=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_OBJ_SUPPORT=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS=y
+CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL=y
+CONFIG_LWM2M_VERSION_1_1=y
+CONFIG_BASE64=y
+CONFIG_LWM2M_RW_SENML_JSON_SUPPORT=y
+CONFIG_LWM2M_RW_JSON_SUPPORT=y
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/overlay-assist-cell.conf:8]: Same questions as with the previous overlay file.

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 20]
```diff
@@ -100,6 +100,39 @@ config LWM2M_CLIENT_UTILS_SIGNAL_MEAS_INFO_INSTANCE_COUNT
...
+
+choice
+	default LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL
+	prompt "Select which method is used for device location"
+
+config LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL
+	bool "Use location based on current cell and possibly neighboring cells"
+
+config LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS
+	bool "Use location based on A-GPS"
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:126]: Using words 'assisted' and 'predicted' here wouldn't hurt.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+#define LOCATION_ASSIST_MAX_ID				12
+
+static int32_t assist_type;
+static int32_t agps_mask;
+static int32_t pgps_pred_count;
+static int32_t pgps_pred_interval;
+static int32_t pgps_start_gps_day;
+static int32_t pgps_start_gps_time_of_day;
+#if defined(CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS)
+static char assist_buf[4096];
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/location_assistance_obj.c:58]: Should these be configurable?  Or at least set through the Kconfig and provide some reasoning for the chosen values even though there wouldn't be a need to change those.

**Review by VeijoPesonen — APPROVED** [review]
Just minor notes, probably ignorable.

**Comment by SeppoTakalo** [comment]
Rebased to retrigger the tests.

File: samples/nrf9160/lwm2m_client/overlay-assist-cell.conf (hunk 19 above)
Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/overlay-assist-cell.conf:8]: OLd JSON shuold remove

**Pull request merged**